eth/eth2.go

```diff
@@ -0,0 +1,325 @@
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package eth
+
+import (
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	chainParams "github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/trie"
+)
+
+type Eth2API struct {
+	eth  *Ethereum
+	env  *eth2bpenv
+	head common.Hash
+}
+
+// NewEth2API creates a new API definition for the eth2 prototype.
+func NewEth2API(eth *Ethereum) *Eth2API {
+	return &Eth2API{eth: eth}
+}
+
+type eth2bpenv struct {
+	state   *state.StateDB
+	tcount  int
+	gasPool *core.GasPool
+
+	header   *types.Header
+	txs      []*types.Transaction
+	receipts []*types.Receipt
+}
+
+func (api *Eth2API) commitTransaction(tx *types.Transaction, coinbase common.Address) error {
+	//snap := eth2rpc.current.state.Snapshot()
+
+	chain := api.eth.BlockChain()
+	receipt, err := core.ApplyTransaction(chain.Config(), chain, &coinbase, api.env.gasPool, api.env.state, api.env.header, tx, &api.env.header.GasUsed, *chain.GetVMConfig())
+	if err != nil {
+		//w.current.state.RevertToSnapshot(snap)
+		return err
+	}
+	api.env.txs = append(api.env.txs, tx)
+	api.env.receipts = append(api.env.receipts, receipt)
+
+	return nil
+}
+
+func (api *Eth2API) makeEnv(parent *types.Block, header *types.Header) error {
+	state, err := api.eth.BlockChain().StateAt(parent.Root())
+	if err != nil {
+		return err
+	}
+	api.env = &eth2bpenv{
+		state:   state,
+		header:  header,
+		gasPool: new(core.GasPool).AddGas(header.GasLimit),
+	}
+	return nil
+}
+
+// Structure described at https://hackmd.io/T9x2mMA4S7us8tJwEB3FDQ
+type AssembleBlockParams struct {
+	ParentHash common.Hash `json:"parent_hash"`
+	Timestamp  uint64      `json:"timestamp"`
+}
+
+// Structure described at https://ethresear.ch/t/executable-beacon-chain/8271
+type ExecutableData struct {
+	Coinbase     common.Address       `json:"coinbase"`
+	StateRoot    common.Hash          `json:"state_root"`
+	GasLimit     uint64               `json:"gas_limit"`
+	GasUsed      uint64               `json:"gas_used"`
+	Transactions []*types.Transaction `json:"transactions"`
+	ReceiptRoot  common.Hash          `json:"receipt_root"`
+	LogsBloom    []byte               `json:"logs_bloom"`
+	BlockHash    common.Hash          `json:"block_hash"`
+	ParentHash   common.Hash          `json:"parent_hash"`
+	Difficulty   *big.Int             `json:"difficulty"`
+}
```

Comment:
There is a slightly different field set for the `ExecutionPayload` in the Rayonism spec. Please, take a look here https://notes.ethereum.org/@n0ble/rayonism-the-merge-spec#Parameters1.